### DISCUSSION TITLE
Package maintainers?

### DISCUSSION THREAD

**Discussion by MarcelWilson** [django-commons/axe-selenium-python#207] [General]
Is this project actively being maintained?  I see a couple PRs submitted from last year without any movement.

**Comment by m8ttyB** [comment]
@kimberlythegeek / @davehunt - is Moz still maintaining this or has the project been sunsetted?

**Comment by m8ttyB** [comment]
Alternatively, can this project be shifted to a community ownership model? Happy to re-take the reigns.

**Comment by MarcelWilson** [comment]
I ended up forking the project as an alternatively named package primarily for myself but went ahead and published it to pypi in case it's useful for anyone else.    ([selenium-axe-python](https://pypi.org/project/selenium-axe-python/))

**Comment by m8ttyB** [comment]
I was the architect of axe-selenium-python and core maintainer. I lost access after leaving Mozilla. I'll contact moz via other channels to understand their intent with this project.

**Comment by thibaudcolas** [comment]
For what it’s worth, we’ve been investigating using this package for [Django](https://github.com/django/django) ([django#16372](https://github.com/django/django/pull/16372)), so I’d love to know what the maintenance status is as well – and help if help is needed.

**Comment by m8ttyB** [comment]
Mozilla staff are attending an All Hands meeting next week. I don't expect to hear back until after their meetings. The gears are in motion, we should receive a status update shortly. I appreciate everyone's patience while we get this sorted out.

**Comment by davehunt** [comment]
I can confirm that this project is no longer being maintained by anyone at Mozilla. I fully support @m8ttyB's suggestion of handing this over to a new owner or community for future development. My suggestion would be to initiate a [repository transfer](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository). If others are in agreement, I believe we would just need to know the new owner.

**Comment by m8ttyB** [comment]
thx @davehunt. I'm working to track down @kimberlythegeek to discuss the [pypi maintainers list](https://pypi.org/project/axe-selenium-python/).

An open question to the community - @MarcelWilson + @thibaudcolas:
[Deque's Python bindings](https://docs.deque.com/devtools-html/4.0.0/en/py-basic-api-example) appear to have significantly improved over the years. The initial goals of this project were to provide Python bindings for axe-core, where Deque (if I remember correctly) only supported JS at the time. This gap has narrowed quite a bit.

Is there an intrinsic reason to continue using `axe-selenium-python` instead of Deque's `axe-devtools-api`? I'm not against reviving this project, but I want a sanity check.

**Comment by m8ttyB** [comment]
I want to nudge this thread again -- I have not been able to locate @kimberlythegeek yet. I'll continue to work on getting ahold of her. In the meantime, @MarcelWilson or @thibaudcolas, have you given my question some thought? Should we suggest the community begin using `axe-devtools-api` instead of this project?

**Comment by thibaudcolas** [comment]
@m8ttyB thanks for following up! I’ve not come across Deque’s `axe-devtools-api` before. It seems to be a different product? [Axe DevTools](https://www.deque.com/axe/devtools/) doesn’t seem to be open source / is only available on an authenticated packages registry. [`axe-core`](https://github.com/dequelabs/axe-core) is what I’m looking for Python/Selenium bindings for. Deque seems to have a lot of language-specific axe-core bindings these days but not one for Python/Selenium. There is a [third-party axe-core-python](https://github.com/ruslan-rv-ua/axe-core-python) though, which advertises Selenium and Playwright.

As far as ownership, the options I can see working from my side are:

- https://github.com/django if there’s agreement from Django ops it makes sense to have a package on there that’s not just for Django use.
- A new dedicated organisation
- A personal account as a last resort (not keen on the bus factor)

We’d then try to replicate this as appropriate in PyPI, set up an Organization account and invite the relevant maintainers.

**Comment by thibaudcolas** [comment]
Hey all, my turn to nudge :)

**Comment by m8ttyB** [comment]
Hi @thibaudcolas -- I was hoping @MarcelWilson might have a moment to weigh in. 

You make fine points, particularly around being pushed towards an authenticated package registry that ties usage to Deque. Is this a blocker? The underlying integration they use to inject Axe with Selenium appears curiously similar to our solution.

Again, I am not dismissing the value of `axe-selenium-python`; I just want to be clear that folks are comfortable creating/maintaining a solution that roughly parallel to Deque's that still relies on `axe-core`. Reinvigorating community engagement on this project would give us more fine-grained control and move it out of the pay-walled model that Deque's model suggests.

If you're a 👍🏼, let's move forward, can you approach the Django project to engage in a conversation and take a temperature check? I agree with your assessment. Once it's decided which organization to host this project under, Mozilla folks will assist with transferring the project - ty to @davehunt & @pdehaan.

 * https://github.com/django if there’s agreement from Django ops, it makes sense to have a package on there that’s not just for Django use.
* A new dedicated organization

> We’d then try to replicate this as appropriate in PyPI, set up an Organization account and invite the relevant maintainers.

There is a mechanism in place to recover abandoned pypi namespaces. I will begin the process.

**Comment by MarcelWilson** [comment]
@m8ttyB I'm not against paid libraries, but unfortunately it would be a non-starter for my projects (one of which is also  open source).   

Personally I would love to see `axe-selenium-python` continue to thrive.  It already seems to have roots, just needs a little TLC.  (it would also mean I can ditch my forked version 😆)

**Comment by m8ttyB** [comment]
ty @MarcelWilson. Let's do it.

- [ ] @thibaudcolas, if you will, please approach the Django community about hosting this project.
- [ ] @m8ttyB will work to transition the pypi project to a state where new maintainers can be added, removed, or deprecated.

**Comment by tushar5526** [comment]
I love the efforts everyone is making here to push this forward. I would be happy to volunteer to help in any manner possible and want to contribute some features inspired by https://github.com/pamelafox/pytest-axe-playwright-snapshot, majorly interested in having inbuilt snapshot testing as starters.

**Comment by thibaudcolas** [comment]
Quick update – took me a while but this is now in front of the right Django people for review. I’ll keep you all posted.

**Comment by m8ttyB** [comment]
Brilliant, ty. That's great news.

**Comment by thibaudcolas** [comment]
Hi folks, I have good and bad news. The bad news is we can’t get this under the [django](https://github.com/django) organization in GitHub, because of the concerns I mentioned earlier about there being a package on there that’s not Django-related enough. The good news is – we can still proceed with the other options I mentioned :)

> - A new dedicated organisation
> - A personal account as a last resort (not keen on the bus factor)

Now I’m not keen on the personal account because of the bus factor – so I think the "separate organisation" is the way to go. I’m happy to set that up if that works for others here. Over the last few months I’ve had a good experience moving [django-recaptcha](https://github.com/django-recaptcha/django-recaptcha) to this type of model.

@m8ttyB what do you think? If this works for you, you can initiate the transfer to me and I’ll take it from there, coordinating with other volunteer maintainers to set up the package in a separate org. And then go through the needed maintenance.

For reference, here are people affiliated with Django who have offered to help maintain this: @tushar5526 @thibaudcolas @knyghty @sabderemane. And @MarcelWilson if you’re up for it, the more the merrier.

**Comment by MarcelWilson** [comment]
I agree a separate github org would make the most sense.  

One **very** minor correction, I'm not affiliated with Django.  I use the axe in conjunction with selenium.  

I mention this because I feel it's important that the purpose of the package isn't/shouldn't be specifically limited to one umbrella.

**Comment by thibaudcolas** [comment]
@m8ttyB 👋 checking in where we’re at with this – do you think we could make the GitHub + PyPI transfer happen? I’m happy to have the package transferred to my personal account and take it from there, or you can create a new separate organization yourself and add the above people to it.

Re the PyPI package, same. You mentioned a mechanism to recover packages there, I’m not aware of it but am happy to take care of that if needed.

**Comment by m8ttyB** [comment]
Oi vey, I apologize. This conversation fell off my map, and messages were sent to `dev/null.`

- Let's get this back on rails and transfer this project to you. What's the right workflow? Transfer it to your personal account, from which you can move it to an organization to negate the bus factor.
- I'll find the docs for PyPI package ownership.

@davehunt, can you help Mozilla formalize the transfer of `axe-selenium-python` to @thibaudcolas? 

/cc @pdehaan

**Comment by m8ttyB** [comment]
I've filed a [support request with PyPI](https://github.com/pypi/support/issues/4205) to initiate the transfer of the maintainer role. Once completed, we'll coordinate a community maintainer model for the package.

**Comment by davehunt** [comment]
I've [opened a ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1902634) for transferring this repository to @thibaudcolas.

**Comment by thibaudcolas** [comment]
Excellent! :) Yes, as soon as the transfer is done I’ll proceed with creating an org for it, with the [people listed above](https://github.com/mozilla-services/axe-selenium-python/issues/192#issuecomment-1936997374) sharing access.

**Comment by davehunt** [comment]
There's some process involved in transferring a repository, which makes me wonder if it's worth the effort. If we're able to get access to the PyPI project, could we archive this repository and allow a fork to become the canonical repository? Apparently this approach has been used before. What do you think?

**Comment by thibaudcolas** [comment]
@davehunt with the caveat that I don’t have context on what your process involves, nor on how PyPI package transfers work, that seems ok to me.

Here’s what I think we would lose, in rough order of value:

1. Automated GitHub redirects from `/mozilla-services/axe-selenium-python`
2. Quick access to the GitHub issues and pull requests history via `#<number>` mentions
3. Watch status at the repository and individual issue / PR level.
4. Connection to the project’s forks
5. GitHub Stars

I’d like to see mitigations in place for items 1. and 2. if possible. I’d run an automated script to recreate placeholder issues for all existing issues, with a link pointing to their source. And it’d be nice if on the existing repository, I could comment with a link to the new repo before it’s archived. And signposting on the README before archival.

---

Be great if we could rely on GitHub’s support for all of this, but if not, I’ve created [axe-selenium-python/axe-selenium-python-fork](https://github.com/axe-selenium-python/axe-selenium-python-fork) for now as a placeholder, which I’ll rename if we proceed with this.

**Comment by davehunt** [comment]
Repository transfer has been approved and should happen soon. Hopefully we can also get some attention on https://github.com/pypi/support/issues/4205 soon. Thank you everyone for your patience.

**Comment by thibaudcolas** [comment]
Hi everyone, there’s been lots of progress! Thanks to @davehunt, the repository transfer is now completed. The repository moved to `thibaudcolas/axe-selenium-python`, and I’ve subsequently moved it to `axe-selenium-python/axe-selenium-python` so we can have @knyghty as an Owner just in case. We might not leave it there, our "plan A" is currently to move the package to [Django Commons](https://github.com/django-commons) which is a new community-driven package maintenance organization.

For transparency, here’s the legal agreement between Mozilla and myself about the repository transfer: [Mozilla Thibaud Axe-Selenium-Python Agreement For Execution (PDF, 200kb)](https://drive.google.com/file/d/1YM5in5Mk2IzqYEGyd6v2zytJC89EYXO8/view?usp=sharing).

## PyPI package

The next step is the PyPI package, for which @m8ttyB opened a PEP 541 request with PyPI back in June ([pypi/support#4205](https://github.com/pypi/support/issues/4205)). Looking at how those requests are processed, the PyPI support team is currently reviewing requests from _March-April 2023_, about 18 months ago. If those delays hold, we’d be getting support in December 2025 / January 2026.

In the meantime, I’ve set up [axe-selenium](https://pypi.org/project/axe-selenium/) on PyPI. **Please don’t switch to it now**. I’ve tried to re-publish the artifacts of [axe-selenium-python 2.1.6](https://pypi.org/project/axe-selenium-python/2.1.6/) as-is, but made no attempt to test the results. This is just there as a placeholder so we have an available package name to publish to.

## Onwards

If anyone following this thread has concerns with Django Commons, please share them now before we proceed? And if anyone is interested to help with future maintenance and hasn’t said so yet, now’s also the right time.

And as far as the implementation – I think we’re in a good place to proceed with adoption of modern Python packaging practices / tools; other maintenance-related project health improvements; and subsequently improvements to the functionality (in particular Axe version upgrades). For now I’ve set up a kanban board: [2024 maintenance overhaul](https://github.com/orgs/axe-selenium-python/projects/1/views/1).

**Comment by tushar5526** [comment]
This is awesome, thanks for all the work!

**Comment by MatMoore** [comment]
Thank you for making this happen. If you need additional maintainers I'd be happy to get involved in the project health improvements.

**Comment by MarcelWilson** [comment]
Just checking in.  Looking at https://github.com/pypi/support/issues/4205 it would appear they couldn't get ahold of "the owner".  Do we have any idea what kind of timeline we can expect?  (assuming they move forward). 

Should we attempt pushing updates to https://pypi.org/project/axe-selenium/ in the interim?

**Comment by thibaudcolas** [comment]
We got access yesterday! 🌈 [axe-selenium-python on PyPI](https://pypi.org/project/axe-selenium-python/) now has me as the owner and I’ve invited @knyghty as well to have some redundancy in the short term.

I’m not super clear on the order of actions but the plan as I see it is still:

- Get a new release out that’s the same as the current latest, with updated metadata to make it clear "new owners"
- New package tooling with a focus on automation
- Move to Django Commons
- Axe upgrade
- (New release with Axe upgrade(s))
- Other feature development
- Onboarding of a sustainable amount of maintainers
- Triage of all existing issues and creation of an informal roadmap

I’ll review the Django Commons process a bit more to check if it makes more sense to add everyone to the package "now" or after the move there.

**Comment by thibaudcolas** [comment]
👋 I’ve now moved this from issues to discussions, reflecting that this is more or less "solved" / the remaining steps can be tracked elsewhere. Quick update of the relevant changes:

- @knyghty is now set up as Owner in PyPI as well as me
- We’re pursuing the move to Django Commons in #205
- Had lots of nice tooling improvements in #196

I believe our next steps are completing that move to Django Commons, Trusted Publishing and the Axe upgrade. Onwards!
